In [ ]:
import numpy as np
import pandas as pd

# **1.File train**

In [ ]:
#Vì file quá lớn nên nhóm em up lên drive rồi upload lên đây, link drive: https://drive.google.com/file/d/1LPzdw8dvC6m5YgTIy9M3lS52LyvvG7VI/view?usp=sharing
!gdown --id 1LPzdw8dvC6m5YgTIy9M3lS52LyvvG7VI


Downloading...
From: https://drive.google.com/uc?id=1LPzdw8dvC6m5YgTIy9M3lS52LyvvG7VI
To: /content/ALL.json
173MB [00:02, 81.2MB/s]


In [ ]:
#Cell này dùng để đưa data về dạng csv và bỏ qua một số hàng bị lỗi khi crawl (không đáng kể)
#Các hàng bị lỗi được in 
f = open('ALL.json','r')
lines = f.readlines()
f1 = open('train.txt','w')
for line in lines:
  try:
    a = line
    if a[-1] == '\n':
      if a[-3] != '0' and a[-3] != '1':
        print(a)
      else:
        f1.write(a[18:a.find('", "headline": "')] + '@_@' + a[a.find('", "headline": "') + 16:-22] + '@_@' + a[-3] + '\n')
    else:
      if a[-3] != '0' and a[-3] != '1':
        print(a)
      else:
        f1.write(a[18:a.find('", "headline": "')] + '@_@' + a[a.find('", "headline": "') + 16:-22] + '@_@' + a[-2] + '\n')
  except:
    print(a)
f.close()
f1.close()

{"article_link": "https://www.thetimes.co.uk/article/adu-lts-to-get-ri-gh-t-to-choose-gender-plwqq6rw8", "headline": "adu



lts to get 

ri

gh

{"article_link": "https://www.thetimes.co.uk/article/carroll-service-w6b09kc69", "headline": "carroll

{"article_link": "https://www.thetimes.co.uk/article/key-br-it-is-h-role-in-iraqi-vi-ct-o-ry-wbclc95x2", "headline": "key br

it

is

h  role in iraqi vi

ct

o

{"article_link": "https://www.thetimes.co.uk/article/wives-at-war-over-dis-graced-dan-czuk-vkm8zdscc", "headline": "wives at war over dis

graced dan

{"article_link": "https://www.thetimes.co.uk/article/ya-rd-ap-ol-og-ises-to-whistlebl-ow-er-s-5rrgpqx99", "headline": "ya

rd ap

ol

og

ises to whistlebl

ow

er

{"article_link": "https://www.thetimes.co.uk/article/mi5-tried-to-recruit-terror-suspect-before-his-flight-to-syria-sh0xjpz6v", "headline": "mi5 tried  to recruit 

{"article_link": "https://www.thetimes.co.uk/article/mps-back-jerusalem-as-england-anthem-m5zf7nwg2", "headl

In [ ]:
# Đọc file CSV
data_train = pd.read_csv("train.txt", delimiter='@_@', names=['article_link','headline',  'is_sarcastic'])
#kiểm tra thông tin của file
data_train

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  


article_link  ... is_sarcastic
0       https://www.thetimes.co.uk/article/aston-villa...  ...            0
1       https://www.thetimes.co.uk/article/antarctic-a...  ...            0
2       https://www.thetimes.co.uk/article/argentina-c...  ...            0
3       https://www.thetimes.co.uk/article/amazon-pain...  ...            0
4       https://www.thetimes.co.uk/article/autumn-as-t...  ...            0
...                                                   ...  ...          ...
842482  https://babylonbee.com/news/zuckerberg-begins-...  ...            1
842483  https://babylonbee.com/news/zuckerberg-emerges...  ...            1
842484  https://babylonbee.com/news/zuckerberg-loses-c...  ...            1
842485  https://babylonbee.com/news/zuckerberg-promise...  ...            1
842486  https://babylonbee.com/news/zuckerberg-require...  ...            1

[842487 rows x 3 columns]

# **2.File test**
*   Để quá trình train diễn ra nhanh chóng và thuận tiện, nhóm em đã hỏi ý và xin data nhóm bạn: Cao Đức trí (link github: https://github.com/hauphanlvc/CS114.L21AndCS114.L22.KHCL) để làm data test.
*   Data test gồm:
    
    breitbart(0): 65996 headlines

    newthump(1): 3805 headlines






In [ ]:
#link drive file tổng hợp 2 trang báo: https://drive.google.com/file/d/1jtl0JpkdVfH402cesMcchrCvmvZTiWzc/view?usp=sharing
!gdown --id 1jtl0JpkdVfH402cesMcchrCvmvZTiWzc

Downloading...
From: https://drive.google.com/uc?id=1jtl0JpkdVfH402cesMcchrCvmvZTiWzc
To: /content/test.json
17.3MB [00:00, 65.3MB/s]


In [ ]:
f = open("test.json", "r")
for x in f:
  # Mỗi dòng định dang lại theo format: *article_link* " *headline* " *is_sarcastic*
  x=x.replace('article_link', "")
  x=x.replace('headline', "")
  x=x.replace('is_sarcastic', "")
  x=x.replace('"": "', "")
  x=x.replace('"": ', "")
  x=x.replace('{', "")
  x=x.replace('}', "")
  x=x.replace('",', '"')
  
  #tạo file test.txt rồi lưu mỗi dòng đã xử lý vào file
  fileOutput = "test.txt"
  f = open(fileOutput, "a", encoding="utf-8")
  f.writelines(x)
  f.close()
  

In [ ]:
#Cell này dùng để đưa data về dạng csv và bỏ qua một số hàng bị lỗi khi crawl (không đáng kể)
#Các hàng bị lỗi được in 
f = open('test.json','r')
lines = f.readlines()
f1 = open('test.txt','w')
for line in lines:
  try:
    a = line
    if a[-1] == '\n':
      if a[-3] != '0' and a[-3] != '1':
        print(a)
      else:
        f1.write(a[18:a.find('", "headline": "')] + '@_@' + a[a.find('", "headline": "') + 16:-22] + '@_@' + a[-3] + '\n')
    else:
      if a[-3] != '0' and a[-3] != '1':
        print(a)
      else:
        f1.write(a[18:a.find('", "headline": "')] + '@_@' + a[a.find('", "headline": "') + 16:-22] + '@_@' + a[-2] + '\n')
  except:
    print(a)
f.close()
f1.close()

In [ ]:
# Đọc file CSV
data_test = pd.read_csv("test.txt", delimiter='@_@', names=['article_link','headline',  'is_sarcastic'])
#kiểm tra thông tin của file
data_test

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  


article_link  ... is_sarcastic
0      https://www.breitbart.com/europe/2021/06/12/wa...  ...            0
1      https://www.breitbart.com/sports/2021/06/12/l-...  ...            0
2      https://www.breitbart.com/politics/2021/06/12/...  ...            0
3      https://www.breitbart.com/politics/2021/06/12/...  ...            0
4      https://www.breitbart.com/border/2021/06/12/10...  ...            0
...                                                  ...  ...          ...
67144  https://newsthump.com/2019/10/03/climate-chang...  ...            1
67145  https://newsthump.com/2019/06/07/tory-leadersh...  ...            1
67146  https://newsthump.com/2019/03/21/im-on-your-si...  ...            1
67147  https://newsthump.com/2019/12/18/whirlpools-ne...  ...            1
67148  https://newsthump.com/2019/06/22/boris-johnson...  ...            1

[67149 rows x 3 columns]

In [ ]:
# link drive của file trang châm biến newsthump: https://drive.google.com/file/d/1ZqSwqFUKrCHQbLwpSvzLtAku3m8j_IWg/view?usp=sharing

In [ ]:
!gdown --id 1ZqSwqFUKrCHQbLwpSvzLtAku3m8j_IWg

Downloading...
From: https://drive.google.com/uc?id=1ZqSwqFUKrCHQbLwpSvzLtAku3m8j_IWg
To: /content/NewsThump.json
100% 975k/975k [00:00<00:00, 52.5MB/s]


In [ ]:
#Cell này dùng để đưa data về dạng csv và bỏ qua một số hàng bị lỗi khi crawl (không đáng kể)
#Các hàng bị lỗi được in 
f = open('NewsThump.json','r')
lines = f.readlines()
f1 = open('NewsThumptest.txt','w')
for line in lines:
  try:
    a = line
    if a[-1] == '\n':
      if a[-3] != '0' and a[-3] != '1':
        print(a)
      else:
        f1.write(a[18:a.find('", "headline": "')] + '@_@' + a[a.find('", "headline": "') + 16:-22] + '@_@' + a[-3] + '\n')
    else:
      if a[-3] != '0' and a[-3] != '1':
        print(a)
      else:
        f1.write(a[18:a.find('", "headline": "')] + '@_@' + a[a.find('", "headline": "') + 16:-22] + '@_@' + a[-2] + '\n')
  except:
    print(a)
f.close()
f1.close()

In [ ]:
# Đọc file CSV
data_test_NT = pd.read_csv("NewsThumptest.txt", delimiter='@_@', names=['article_link','headline',  'is_sarcastic'])
#kiểm tra thông tin của file
data_test_NT

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  


article_link  ... is_sarcastic
0     https://newsthump.com/2021/04/30/larry-the-cat...  ...            1
1     https://newsthump.com/2021/04/19/queen-and-oth...  ...            1
2     https://newsthump.com/2020/02/13/boris-johnson...  ...            1
3     https://newsthump.com/2020/02/01/britain-alrea...  ...            1
4     https://newsthump.com/2019/08/28/queen-brillia...  ...            1
...                                                 ...  ...          ...
3800  https://newsthump.com/2019/10/03/climate-chang...  ...            1
3801  https://newsthump.com/2019/06/07/tory-leadersh...  ...            1
3802  https://newsthump.com/2019/03/21/im-on-your-si...  ...            1
3803  https://newsthump.com/2019/12/18/whirlpools-ne...  ...            1
3804  https://newsthump.com/2019/06/22/boris-johnson...  ...            1

[3805 rows x 3 columns]

In [ ]:
# Data train
#Xóa dòng có giá trị NaN xa khỏi data
data_train = data_train.dropna(how = 'any')
# Lấy 2 cột cần thiết là headline và is_sarcastic
new_data_train = data_train.copy()
new_data_train = new_data_train[['headline', 'is_sarcastic']]

# Data test
#Xóa dòng có giá trị NaN xa khỏi data
data_test = data_test.dropna(how = 'any')
# Lấy 2 cột cần thiết là headline và is_sarcastic
new_data_test = data_test.copy()
new_data_test = new_data_test[['headline', 'is_sarcastic']]

# Data test NewsThump
#Xóa dòng có giá trị NaN xa khỏi data
data_test_NT = data_test_NT.dropna(how = 'any')
# Lấy 2 cột cần thiết là headline và is_sarcastic
new_data_test_NT = data_test_NT.copy()
new_data_test_NT = new_data_test_NT[['headline', 'is_sarcastic']]


In [ ]:
import re 
def format(x):
    #chuyển tất cả sang string
    x=str(x)
    # Loại bỏ các kí tự khác 0 - 9, a - z và khoảng cách
    x = re.sub("[^0-9a-z ]", '', x)
    return x

In [ ]:

new_data_train['headline'] = new_data_train['headline'].apply(format)

new_data_train

headline  is_sarcastic
0       aston villa match postponed after covid19 outb...             0
1       antarctic architect hugh broughton wants warm ...             0
2       argentina captain pablo matera sacked over rac...             0
3            amazon paintings reveal ice age civilisation             0
4                     autumn as told by a farmer and poet             0
...                                                   ...           ...
842482  zuckerberg begins testimony by reminding senat...             1
842483  mark zuckerberg emerges from under mans bed to...             1
842484  zuckerberg loses contact lens during senate he...             1
842485  zuckerberg promises new dating service will re...             1
842486  zuckerberg requires house members to accept te...             1

[842471 rows x 2 columns]

In [ ]:

new_data_test['headline'] = new_data_test['headline'].apply(format)

new_data_test

headline  is_sarcastic
0          rench areft eader oused with lour by rotester             0
1       uper owl ost ommittee ives rant to ctivists e...             0
2       alling hort of 70 ully accinated ask andate t...             0
3                        426 of  dults re ully accinated             0
4       100 igrants ound in tash ouse near order in exas             0
...                                                  ...           ...
67144  climate change u2018written off as a hoaxu2019...             1
67145  tory leadership contest to introduce swimsuit ...             1
67146  u201ciu2019m on your sideu201d insists woman w...             1
67147  whirlpoolu2019s new christmas advert shows fam...             1
67148  boris johnson wakes from massive bender hoping...             1

[67149 rows x 2 columns]

In [ ]:

new_data_test_NT['headline'] = new_data_test_NT['headline'].apply(format)

new_data_test_NT

headline  is_sarcastic
0     larry the cat already scratched the shit out o...             1
1     queen and other top monarchs to break away and...             1
2                boris johnson unveils new look cabinet             1
3     britain already 36 more british since leaving ...             1
4     queen brilliantly steers attention away from p...             1
...                                                 ...           ...
3800  climate change u2018written off as a hoaxu2019...             1
3801  tory leadership contest to introduce swimsuit ...             1
3802  u201ciu2019m on your sideu201d insists woman w...             1
3803  whirlpoolu2019s new christmas advert shows fam...             1
3804  boris johnson wakes from massive bender hoping...             1

[3805 rows x 2 columns]

# **3.Removing stop words with NLTK in Python**


In [ ]:
#Nguồn tham khảo: https://www.geeksforgeeks.org/removing-stop-words-nltk-python/
import nltk 
nltk.download('punkt')
from nltk.tokenize import word_tokenize
nltk.download("stopwords")
from nltk.corpus import stopwords
x_train_nltk = []
for s in new_data_train['headline']:
  s=str(s)
  x_train_nltk.append(word_tokenize(s))

x_test_nltk = []
for s in new_data_test['headline']:
  s=str(s)
  x_test_nltk.append(word_tokenize(s))

x_test_NT_nltk = []
for s in new_data_test_NT['headline']:
  s=str(s)
  x_test_NT_nltk.append(word_tokenize(s))

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
#danh sách tất cả stop words
all_stopwords=set(stopwords.words('english'))
print(all_stopwords)
print(len(all_stopwords))

{'again', 'isn', 'didn', 'yourself', 'nor', 'or', 'what', 'does', 'i', "you'll", 'the', 'do', "aren't", 'were', 't', "won't", "couldn't", 'same', 'which', 'your', 'was', 'so', 'but', 'hers', 'between', 'm', 've', 'no', "haven't", "needn't", "didn't", 'been', 'its', 'ma', "doesn't", 'we', 'to', 'doing', 'up', 'now', 'did', 'wasn', 'them', 'this', 'more', 'each', 'below', "mightn't", 'mustn', 'when', 'than', 'from', 'him', 'me', 'being', 'in', 'for', "weren't", 'needn', 'ourselves', 'theirs', 'under', 'you', 'yourselves', 'once', 'both', 'before', 'haven', 'ours', "she's", 'don', 'our', 'her', 'other', 'whom', 'couldn', 'y', 'it', 'over', "it's", 'against', 'because', 'until', "don't", 'those', 'while', "you're", "that'll", 'into', "you'd", 'at', 'has', 'few', 'having', 'some', 'there', 'his', 's', 'herself', 'she', 'by', 'mightn', 'shan', 'be', 'then', 'through', 'not', 'with', 'myself', 'and', 'have', 'own', 'weren', 'about', 'on', 'wouldn', 're', 'ain', 'am', 'hasn', "you've", 'as', '

In [ ]:
#loại bỏ stop words trong data train:
filtered_sentence_train = []
for words in x_train_nltk:
  filtered_sentence=[]
  for w in words:
    if w not in all_stopwords:
        filtered_sentence.append(w)
  filtered_sentence_train.append(filtered_sentence)
#loại bỏ stop words trong data test:
filtered_sentence_test = []
for words in x_test_nltk:
  filtered_sentence=[]
  for w in words:
    if w not in all_stopwords:
        filtered_sentence.append(w)
  filtered_sentence_test.append(filtered_sentence)

#loại bỏ stop words trong data test NewsThump:
filtered_sentence_test_NT = []
for words in x_test_NT_nltk:
  filtered_sentence=[]
  for w in words:
    if w not in all_stopwords:
        filtered_sentence.append(w)
  filtered_sentence_test_NT.append(filtered_sentence)

# **4.Train với data chưa xóa stop_word**

In [ ]:
x_train, y_train = np.array(new_data_train['headline']), np.array(new_data_train['is_sarcastic'])
x_test, y_test = np.array(new_data_test['headline']), np.array(new_data_test['is_sarcastic'])
x_test_NT, y_test_NT= np.array(new_data_test_NT['headline']), np.array(new_data_test_NT['is_sarcastic'])

In [ ]:
#Nguồn code tham khảo: https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
#                      https://towardsdatascience.com/machine-learning-nlp-text-classification-using-scikit-learn-python-and-nltk-c52b92a7c73a
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
import numpy as np

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

text_clf.fit(x_train, y_train)
#predict với data tổng hợp
y_predict = text_clf.predict(x_test)

#Đo độ chính xác
np.mean(y_predict == y_test) 


0.9428733115906417

In [ ]:


#predict chỉ với data trang châm biếm NewsThump
y_predict_NT = text_clf.predict(x_test_NT)

#Đo độ chính xác NT:
np.mean(y_predict_NT == y_test_NT)

0.00026281208935611036

In [ ]:
from sklearn.linear_model import SGDClassifier
text_clf = Pipeline([
     ('vect', CountVectorizer()),
     ('tfidf', TfidfTransformer()),
     ('clf', SGDClassifier()),
 ])
text_clf.fit(x_train, y_train)

y_predict = text_clf.predict(x_test)
#Đo độ chính xác
np.mean(y_predict == y_test) 

0.9433349714813326

In [ ]:
#predict chỉ với data trang châm biếm NewsThump

y_predict_NT = text_clf.predict(x_test_NT)
#Đo độ chính xác
np.mean(y_predict_NT == y_test_NT) 

0.0

# **4.Train với data đã xóa stop_word**

In [ ]:
x_train, y_train = filtered_sentence_train, new_data_train['is_sarcastic']
x_test, y_test =  filtered_sentence_test, new_data_test['is_sarcastic']
x_test_NT, y_test_NT = filtered_sentence_test_NT, new_data_test_NT['is_sarcastic']

In [ ]:
# Vì data train và test sau khi xóa stop words là list không fit được nên chuyển tập list thành string 
# Nguồn code tham khảo: https://www.geeksforgeeks.org/python-program-to-convert-a-list-to-string/
x_train = [" ".join(x) for x in x_train]
x_test =[" ".join(x) for x in x_test]
x_test_NT = [" ".join(x) for x in x_test_NT]

In [ ]:
from sklearn.linear_model import SGDClassifier
text_clf = Pipeline([
     ('vect', CountVectorizer()),
     ('tfidf', TfidfTransformer()),
     ('clf', SGDClassifier()),
 ])
text_clf.fit(x_train, y_train)

y_predict = text_clf.predict(x_test)
#Đo độ chính xác
np.mean(y_predict == y_test) 

0.9433349714813326

In [ ]:
#predict chỉ với data trang châm biếm NewsThump

y_predict_NT = text_clf.predict(x_test_NT)
#Đo độ chính xác
np.mean(y_predict_NT == y_test_NT) 

0.0

In [ ]:
text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

text_clf.fit(x_train, y_train)

y_predict = text_clf.predict(x_test)
#Đo độ chính xác
y_predict
np.mean(y_predict == y_test) 

0.9427541735543344

In [ ]:


#predict chỉ với data trang châm biếm NewsThump
y_predict_NT = text_clf.predict(x_test_NT)

#Đo độ chính xác NT:
np.mean(y_predict_NT == y_test_NT)

0.0005256241787122207

# **5.Điều chỉnh train data train**

In [ ]:
sar_data = new_data_train[new_data_train['is_sarcastic'] == 1] #Lọc các tiêu đề báo châm biếm từ DF ban 
sar_data

headline  is_sarcastic
355262  elon musk admits he wants to travel to mars be...             1
355263  gamer begins wiki deep dive to catch up on ser...             1
355264  mans cell phone disagrees that alexa is listen...             1
355265  gamer captures bigfoot at atrocious 1080p reso...             1
355266  even conspiracy theorist kinda surprised how r...             1
...                                                   ...           ...
842482  zuckerberg begins testimony by reminding senat...             1
842483  mark zuckerberg emerges from under mans bed to...             1
842484  zuckerberg loses contact lens during senate he...             1
842485  zuckerberg promises new dating service will re...             1
842486  zuckerberg requires house members to accept te...             1

[22180 rows x 2 columns]

In [ ]:
nosar_data = new_data_train[new_data_train['is_sarcastic'] == 0].head(len(sar_data)) #Lọc các tiêu đề báo chính thống với số lượng bằng số lượng tiêu đề báo châm 
nosar_data

headline  is_sarcastic
0      aston villa match postponed after covid19 outb...             0
1      antarctic architect hugh broughton wants warm ...             0
2      argentina captain pablo matera sacked over rac...             0
3           amazon paintings reveal ice age civilisation             0
4                    autumn as told by a farmer and poet             0
...                                                  ...           ...
22175                         tv detector vans truth ray             0
22176  times sport unseen silvas farewell to england ...             0
22177     the times view on jimmy lais arrest plucky jim             0
22178  teen flying machine alphonso davies can down l...             0
22179            this country team to make period comedy             0

[22180 rows x 2 columns]

In [ ]:
#Các biến để train,
new_x_train = pd.concat([sar_data['headline'],nosar_data['headline']])
new_y_train = pd.concat([sar_data['is_sarcastic'],nosar_data['is_sarcastic']])
x_test, y_test = np.array(new_data_test['headline']), np.array(new_data_test['is_sarcastic'])
x_test_NT, y_test_NT= np.array(new_data_test_NT['headline']), np.array(new_data_test_NT['is_sarcastic'])

In [ ]:
#Nguồn code tham khảo: https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
#                      https://towardsdatascience.com/machine-learning-nlp-text-classification-using-scikit-learn-python-and-nltk-c52b92a7c73a
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
import numpy as np

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

text_clf.fit(new_x_train,new_y_train)
#predict với data tổng hợp
y_predict = text_clf.predict(x_test)

#Đo độ chính xác
np.mean(y_predict == y_test) 


0.4808708990454065

In [ ]:
#predict chỉ với data trang châm biếm NewsThump
y_predict_NT = text_clf.predict(x_test_NT)

#Đo độ chính xác NT:
np.mean(y_predict_NT == y_test_NT)

0.5897503285151117

* Thêm công đoạn remove stop words

In [ ]:
new_x_train_nltk = []
for s in new_x_train:
  s=str(s)
  new_x_train_nltk.append(word_tokenize(s))

In [ ]:
#loại bỏ stop words trong data train:
new_filtered_sentence_train = []
for words in new_x_train_nltk:
  filtered_sentence=[]
  for w in words:
    if w not in all_stopwords:
        filtered_sentence.append(w)
  new_filtered_sentence_train.append(filtered_sentence)

In [ ]:
#Các biến để train, test
new_x_train= new_filtered_sentence_train
x_test, y_test =  filtered_sentence_test, new_data_test['is_sarcastic']
x_test_NT, y_test_NT = filtered_sentence_test_NT, new_data_test_NT['is_sarcastic']

In [ ]:
x_train = [" ".join(x) for x in new_x_train]
x_test =[" ".join(x) for x in x_test]
x_test_NT = [" ".join(x) for x in x_test_NT]

In [ ]:
#Nguồn code tham khảo: https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
#                      https://towardsdatascience.com/machine-learning-nlp-text-classification-using-scikit-learn-python-and-nltk-c52b92a7c73a
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
import numpy as np

text_clf = Pipeline([
    ('vect', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('clf', MultinomialNB()),
])

text_clf.fit(x_train,new_y_train)
#predict với data tổng hợp
y_predict = text_clf.predict(x_test)

#Đo độ chính xác
np.mean(y_predict == y_test) 


0.5863676301955353

In [ ]:
#predict chỉ với data trang châm biếm NewsThump
y_predict_NT = text_clf.predict(x_test_NT)

#Đo độ chính xác NT:
np.mean(y_predict_NT == y_test_NT)

0.5553219448094613

# **6.Nhận xét và đánh giá**


*   Sau khi nhóm thử thử nghiệm phân loại bằng SGDClassifier và Naive Bayes trước(0) và sau(1) khi remove stop words thì thu được kết quả có độ chính xác: SGDClassifier(0): 94,2912%, SGDClassifier(1): 94,2912%, Naive Bayes(0):  94,2927%, Naive Bayes(1): 94,2942%. Với dataset mà nhóm thu thập được khi test trên 2 trang báo khác, nhìn số liệu trên thì có vẻ kết quả tốt nhưng thực tế do số lượng tiêu đề báo chính thống quá lớn so với tiêu đề báo châm biếm nên model có xu hướng predict ra kết quả là '0'. Mà data dùng để test cũng tương tự vậy, số lượng tiêu đề báo chính thống lớn nên khi predict và đo độ chính xác sẽ ra kết quả như trên.

*   Kết quả này tốt khi predict báo chính thống nhưng tệ khi predict báo châm , bằng chứng là khi predict với data trang châm biếm NewThump thì độ chính xác đều gần bằng hoặc bằng 0%.

*   Hai model SGDClassifier và Naive Bayes làm việc với text data tốt nhưng Data của các trang chính thống quá lớn so với data của những trang châm biến dẫn đến việc train và dự đoán kết quả không cao. Nhóm đi đến dự đoán rằng các "từ" có thể dùng để xác định được trang châm biếm lại xuất hiện nhiều hon ở các tiêu đề báo chính thống do chênh lệch dự liệu quá lớn dẫn tới model predict sai.
Ở phần 5 khi train với lượng data nhỏ hơn nhưng cân bằng giữa số tiêu đề châm biếm và chính thống thì model đã có thể predict đúng được hơn 58% số tiêu đề của NewsThump đổi lại thì khi train chung với báo chính thống độ chính xác đã giảm còn khoảng 48% (khi remove stop words thì độ chính xác lần lượt là 58%, 55% )


